### PR TITLE
Use JAX instead of PyTorch

### PR DESCRIPTION
**Pull request by z0u**
Ported all ML code to use JAX instead of PyTorch for faster iteration on experiments due to less class-based coupling.

PyTorch dependencies have been fully removed. Local (dev) installation only installs the default CPU packages, but GPU deps are included when building Modal images.

### CODE REVIEW

File: src/experiment/model/ngpt.py [hunk 1]
```diff
@@ -68,44 +85,59 @@ def forward(self, x: Float[Tensor, 'B T C'], enc: RotaryEncoding):
...
+        att = (q @ k.swapaxes(-2, -1)) * (self.qk_scale if self.full else self.s_qk())
+        att = jnp.where(jnp.tril(jnp.ones((T, T), bool)), att, -jnp.inf)
+        att_key, out_key = split_keys(key, 2)
+        att = self.dropout(jax.nn.softmax(att, axis=-1), key=att_key)
         y = att @ v
 
         y = merge_heads(y)
-        return self.dropout(self.proj(y))
+        return self.dropout(self.proj(y), key=out_key)
+
```
Comment (z0u) [inline src/experiment/model/ngpt.py:98]: Wait, I thought nGPT didn't use dropout?

Comment (z0u) [reply src/experiment/model/ngpt.py:98]: Oh, it's disabled (configured as 0) in the notebooks. I think we should just remove it from the architecture tho, because it complicates things. Raise a ValueError or similar if dropout is non-zero.
